// cp-lht-spinner/esm2022/lib/lht-spinner.module.mjs
import { NgModule } from '@angular/core';
import { LhtSpinnerComponent } from './component/lht-spinner.component';
import { LhtSpinnerSettingsService } from '../lib/services/lht-spinner-settings.service';
import * as i0 from "@angular/core";
export class LhtSpinnerModule {
    static forRoot(config) {
        return {
            ngModule: LhtSpinnerModule,
            providers: [
                LhtSpinnerSettingsService,
                { provide: 'config', useValue: config },
            ],
        };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtSpinnerModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule }); }
    static { this.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "17.0.4", ngImport: i0, type: LhtSpinnerModule, declarations: [LhtSpinnerComponent], exports: [LhtSpinnerComponent] }); }
    static { this.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtSpinnerModule }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtSpinnerModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [LhtSpinnerComponent],
                    imports: [],
                    exports: [LhtSpinnerComponent],
                }]
        }] });
