// virava/dist/src/utils/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATUS_CUSTOM_LOGIN_ERROR = exports.STATUS_FORBIDDEN = exports.STATUS_UNAUTHORIZED = exports.CONTENT_TYPE_FORM_URL_ENCODED = exports.CONTENT_TYPE_JSON = exports.API_ENDPOINT_RESET_PASSWORD = exports.API_ENDPOINT_REGISTER = exports.API_ENDPOINT_CHANGE_PASSWORD = exports.API_ENDPOINT_TOKEN = exports.API_ENDPOINT_LOGOUT = exports.ACCESS_TOKEN_UPDATE_MIN_VALIDITY = exports.MS_CHECK_TOKEN_VALIDITY_INTERVAL = exports.KEY_EXPIRATION_REFRESH = exports.KEY_EXPIRATION_DATE = exports.KEY_REFRESH_TOKEN = exports.KEY_ACCESS_TOKEN = exports.MS_FOR_ONE_SECOND = exports.GRANT_TYPE_PASSWORD = exports.GRANT_TYPE_REFRESH_TOKEN = void 0;
exports.GRANT_TYPE_REFRESH_TOKEN = 'refresh_token';
exports.GRANT_TYPE_PASSWORD = 'password';
exports.MS_FOR_ONE_SECOND = 1000;
exports.KEY_ACCESS_TOKEN = 'access_token';
exports.KEY_REFRESH_TOKEN = 'refresh_token';
exports.KEY_EXPIRATION_DATE = 'expiration_date';
exports.KEY_EXPIRATION_REFRESH = 'expiration_refresh';
exports.MS_CHECK_TOKEN_VALIDITY_INTERVAL = 10000;
exports.ACCESS_TOKEN_UPDATE_MIN_VALIDITY = 300;
// API
exports.API_ENDPOINT_LOGOUT = '{BASE_URL}/realms/{REALM}/protocol/openid-connect/logout';
exports.API_ENDPOINT_TOKEN = '{BASE_URL}/realms/{REALM}/protocol/openid-connect/token';
exports.API_ENDPOINT_CHANGE_PASSWORD = '/changePassword';
exports.API_ENDPOINT_REGISTER = '/register';
exports.API_ENDPOINT_RESET_PASSWORD = '/resetPassword';
// Content-Type
exports.CONTENT_TYPE_JSON = 'application/json';
exports.CONTENT_TYPE_FORM_URL_ENCODED = 'application/x-www-form-urlencoded';
// Status Codes
exports.STATUS_UNAUTHORIZED = 401;
exports.STATUS_FORBIDDEN = 403;
exports.STATUS_CUSTOM_LOGIN_ERROR = 417;
