// cp-lht-spinner/esm2022/lib/services/lht-spinner-settings.service.mjs
import { Injectable, Inject } from '@angular/core';
import * as i0 from "@angular/core";
export class LhtSpinnerSettingsService {
    constructor(config) {
        this.config = config;
    }
    get libConfig() {
        return this.config;
    }
    setLibConfig(newConfig) {
        this.config = { ...this.config, ...newConfig };
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtSpinnerSettingsService, deps: [{ token: 'config' }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtSpinnerSettingsService }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtSpinnerSettingsService, decorators: [{
            type: Injectable
        }], ctorParameters: () => [{ type: undefined, decorators: [{
                    type: Inject,
                    args: ['config']
                }] }] });
