// cp-lht-spinner/esm2022/lib/services/lht-loading.service.mjs
import { Injectable } from '@angular/core';
import { ComponentPortal } from '@angular/cdk/portal';
import { LhtSpinnerComponent } from '../component/lht-spinner.component';
import * as i0 from "@angular/core";
import * as i1 from "@angular/cdk/overlay";
import * as i2 from "./lht-spinner-settings.service";
export class LhtLoadingService {
    constructor(overlay, lhtSpinnerSettingsService) {
        this.overlay = overlay;
        this.lhtSpinnerSettingsService = lhtSpinnerSettingsService;
        this.overlayRef = null;
        this.requestsCount = 0;
        this.showTimer = null;
    }
    show(spinnerLoadingText) {
        this.requestsCount++;
        const delayTime = this.lhtSpinnerSettingsService.libConfig.spinnerDelayTime || 0;
        if (spinnerLoadingText) {
            this.lhtSpinnerSettingsService.setLibConfig({
                ...this.lhtSpinnerSettingsService.libConfig,
                spinnerLoadingText,
            });
        }
        if (!this.overlayRef) {
            this.overlayRef = this.overlay.create();
        }
        if (this.showTimer === null) {
            if (delayTime > 0) {
                this.showTimer = setTimeout(() => {
                    this.attachSpinner();
                }, delayTime);
            }
            else {
                this.attachSpinner();
            }
        }
    }
    hide() {
        this.requestsCount--;
        if (this.requestsCount <= 0) {
            if (this.showTimer !== null) {
                clearTimeout(this.showTimer);
                this.showTimer = null;
            }
            this.requestsCount = 0;
            if (this.overlayRef && this.overlayRef.hasAttached()) {
                this.overlayRef.detach();
            }
        }
    }
    attachSpinner() {
        if (this.overlayRef && !this.overlayRef.hasAttached()) {
            const loaderPortal = new ComponentPortal(LhtSpinnerComponent);
            this.overlayRef.attach(loaderPortal);
        }
        this.showTimer = null;
    }
    static { this.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtLoadingService, deps: [{ token: i1.Overlay }, { token: i2.LhtSpinnerSettingsService }], target: i0.ɵɵFactoryTarget.Injectable }); }
    static { this.ɵprov = i0.ɵɵngDeclareInjectable({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtLoadingService, providedIn: 'root' }); }
}
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "17.0.4", ngImport: i0, type: LhtLoadingService, decorators: [{
            type: Injectable,
            args: [{
                    providedIn: 'root',
                }]
        }], ctorParameters: () => [{ type: i1.Overlay }, { type: i2.LhtSpinnerSettingsService }] });
